### PR TITLE
[Libra Framework] Proposed handling of errors/usage of new prover features.

### PR DESCRIPTION
**Pull request by wrwg**
This PR introduces a new module [`CoreErrors`](https://github.com/wrwg/libra/blob/errors_module/language/stdlib/modules/CoreErrors.move) which contains functions for global error codes (similar as `CoreAddresses` today). It then uses this module in some places in the framework and applies the new-style abort specifications with codes. Abort specs with codes make only sense with global errors, as they propagate through the stack. This work was started with `LibraTimestamp`, which has [a revamped specification](https://github.com/wrwg/libra/blob/errors_module/language/stdlib/modules/LibraTimestamp.move) following the new methodology. The changes propagated to users of time stamp, of which there are a few.

The approach to global error codes is based on the principle that the set of codes should be small and describe the problem to end users, not the implementation details. This is assuming that debugging can be achieved by other means (replay, VM stack trace).

This work is not complete at all, and need to be continued in subsequent PRs. However, before continuing, we need to agree on the handling of global errors, so this represents a first step as a discussion basis.

Tool changes:
- Making warnings by default off. There is an option `--warn` to turn them on. This helps us for now to quiet warnings about unused schemas, which are needed for the style of spec in this PR.
- There was a bug in the computation of locations for error messages about wrong abort codes. For now, the fancy previous way which was buggy has been turned off.

## Motivation

Consolidate the error codes.

## Test Plan

Existing tests.

## Related PRs

NA

### CODE REVIEW

**Comment by wrwg** [comment]
Added @tzakian and @sblackshear for the Errors module, prover team for entire PR. Open to rename things (Errors -> AbortCodes or whatever). Apologies for the multiple other changes which are not always directly related.

**Comment by wrwg** [comment]
> just a comment given I saw the PR and wanted to pass all my disapproval of constants exposed through functions. I find it not just distasteful, but also heavy and providing absolutely no useful abstraction. There I said it! I am not sure I follow what the errors give you that a pattern of global errors exposed via every single module does not. I understand you could prove that the "error module" returns different values for each function and as such it may give you more guarantees on what gets returned. I am not clear how critical that is to the prover.
> I am also very worried of us adding a module that contains a set of functions for constants for the current code base, which seems to imply that even the deployment of a new single and self containing module may require a WriteSet transaction to add an error.
> I have not reviewed this in depth though so I may be off and apology if that is the case.
> Otherwise, please, don't add a module with functions returning constants. Please....

There is really no way how we can avoid this apart of just forgetting about documenting and specifying *anything* about error codes. 

Let us take as an example `LibraTimestamp::update_global_time`. This function aborts if called from genesis or if not called with the reserved VM account with according error codes. A caller of this function will abort with the same error codes as the abort bubbles up the call stack. If we do not have a global definition of those error codes, we cannot document or specify this. We may even get inconsistencies. For example, the caller of this function may also need to abort if called from genesis by its own means, but produce a different error code. Now you have two different error codes which describe the same situation, namely being called from genesis. Worse, you may even have a the *same* error code for different reasons why the caller aborts, because local error codes have been defined which happen to describe different error situations in two modules.

We have dozens and dozens of similar cases like this example.

I hear your dislake and distaste, and hope it motivates the language team to make constants global eventually. But "distaste" cannot dominate our approach to clean API design. FWIW, we do have the same pattern (nullary functions for constants) today already with `CoreAddresses`; while it is not beautiful, it's actually not a big deal IMO to add the extra `()`.

**Comment by bob-wilson** [comment]
I'm adding @tnowacki as a reviewer here. We really need to get his input on this approach.

**Comment by wrwg** [comment]
> I'm adding @tnowacki as a reviewer here. We really need to get his input on this approach.

Happy to get feedback, but really like to call out that we should keep two things distinct:

1. Methodology to design and specify error codes
2. Language support for them, whether and when

This PR is about (1) and not (2), which I thought we wont tackle before 9/15, based on the discussions we had last week. FWIW, from language viewpoint, this PR uses just the same pattern as the existing `CoreAddresses` module.

**Comment by wrwg** [comment]
Base on a discussion in a prover meeting today including Sam, I added a section about custom error codes to [Errors](https://github.com/wrwg/libra/blob/errors_module/language/stdlib/modules/Errors.move). It's at the end of the linked module. Please check out if you are wondering how to deal with those. 

This also addresses this comment of @dariorussi:

> I am also very worried of us adding a module that contains a set of functions for constants for the current code base, which seems to imply that even the deployment of a new single and self containing module may require a WriteSet transaction to add an error.

The answer to this is that new error codes would be very rarely introduced in the `Errors` module, but as custom codes in the according module. A better name for `Errors` might be `CommonErrors`.

**Comment by sblackshear** [comment]
Some proposed rules of thumb for (1):
A. Error codes are for consumption by clients, not for Libra framework developers or module authors.

B. Every error code should be considered part of the client-facing public API. If the error code doesn't make sense for someone who will read docs, but never look at framework code, it's too low-level. Similarly, every defined error code should be reachable from a current or planned future tx script (or else we should get rid of the error code).

C. Every error code should have an intended resolution. That is: the docs should say (or be able to say) "if you the client receive this error code, here is what you should do in your next transaction to avoid seeing this error code". If the error code doesn't have an intended resolution, it is too low-level and we should report the problem at a different level of abstraction. For example, instead of reporting a generic RESOURCE_NOT_INITIALIZED error code, we should emit an error code specific to the module/resource that tells the user which function/script/args they should use to initialize it.

D. An error code should be defined inside the module that understands how to resolve it. This is typically the same as the module that raises the error code for "leaf" modules (e.g., `LibraAccount::INSUFFICIENT_BALANCE`). But there are also "utility" modules like `Roles` that define many error codes + their resolutions, but do not raise the error codes.

E. A module `A` that `use`s another module `B` should reuse `B`'s error codes for problems that `B` knows how to resolve. E.g., `LibraAccount` should not define its own error code for `NOT_GENESIS`; it should reuse the `Genesis::NOT_GENESIS` error code.

Curious to hear which rules of thumb folks like/don't like and what other ones they have in mind.

**Comment by sblackshear** [comment]
For concreteness, here is how I would apply the principals above to the `Errors` module:

D says:
- `NOT_GENESIS` and `NOT_OPERATING` should be moved to `LibraTimestamp` because it is the module that exposes `is_genesis()`
- `NOT_LIBRA_ROOT_ADDRESS` and `NOT_VM_RESERVED_ADDRESS` should be moved to `CoreAddresses`
- `NOT_LIBRA_ROOT_ROLE` and `NOT_TREASURY_COMPLIANCE_ROLE` should be moved to `Roles`

A/B/C suggests that the remaining error codes might be too generic (but might not be; it's somewhat of a case-by-case thing). For example: let's look at a use of `INVALID_INPUT` in `LibraTimestamp::now_microsconds()`:

```
 public fun now_microseconds(): u64 acquires CurrentTimeMicroseconds {
        assert(exists<CurrentTimeMicroseconds>(CoreAddresses::LIBRA_ROOT_ADDRESS()), Errors::NOT_INITIALIZED());
        borrow_global<CurrentTimeMicroseconds>(CoreAddresses::LIBRA_ROOT_ADDRESS()).microseconds
```

This is not resolvable--the caller can't do anything to publish this resource. But if we instead assert that it's not genesis, that implies that the resource exists. And the caller has a clearer action in the case that it hits that assertion--wait until genesis is over to call this function.

**Comment by dariorussi** [comment]
> > just a comment given I saw the PR and wanted to pass all my disapproval of constants exposed through functions. I find it not just distasteful, but also heavy and providing absolutely no useful abstraction. There I said it! I am not sure I follow what the errors give you that a pattern of global errors exposed via every single module does not. I understand you could prove that the "error module" returns different values for each function and as such it may give you more guarantees on what gets returned. I am not clear how critical that is to the prover.
> > I am also very worried of us adding a module that contains a set of functions for constants for the current code base, which seems to imply that even the deployment of a new single and self containing module may require a WriteSet transaction to add an error.
> > I have not reviewed this in depth though so I may be off and apology if that is the case.
> > Otherwise, please, don't add a module with functions returning constants. Please....
> 
> There is really no way how we can avoid this apart of just forgetting about documenting and specifying _anything_ about error codes.
> 
> Let us take as an example `LibraTimestamp::update_global_time`. This function aborts if called from genesis or if not called with the reserved VM account with according error codes. A caller of this function will abort with the same error codes as the abort bubbles up the call stack. If we do not have a global definition of those error codes, we cannot document or specify this. We may even get inconsistencies. For example, the caller of this function may also need to abort if called from genesis by its own means, but produce a different error code. Now you have two different error codes which describe the same situation, namely being called from genesis. Worse, you may even have a the _same_ error code for different reasons why the caller aborts, because local error codes have been defined which happen to describe different error situations in two modules.
> 
> We have dozens and dozens of similar cases like this example.
> 
> I hear your dislike and distaste, and hope it motivates the language team to make constants global eventually. But "distaste" cannot dominate our approach to clean API design. FWIW, we do have the same pattern (nullary functions for constants) today already with `CoreAddresses`; while it is not beautiful, it's actually not a big deal IMO to add the extra `()`.

I do see the difficulty here. And I think the case where you want to return the same error code from multiple modules is particularly difficult and challenging. I am not sure what could be the right way to solve it.
Again though, updates of modules is a difficult thing now and possibly always. 
The idea that adding an error code requires a WriteSet transaction it is not very appealing. And even if we come up with updates the idea that someone has to override a module because an error code is added seems pretty difficult to accomplish.

I think the point that I am trying to make is that single repository of things in our system does not work too well. Whereas an approach where each module has a chance to be self contained has much less friction.
So, in that respect, I'd like for us all (and you in particular) to think about this and see if there are alternatives that do not require a single point of aggregation (global constants or whatever). If we can get some way to build maps for errors (from a local one to a global one) that the prover could use that would be awesome.
For instance, you could imagine the spec to provide a single global error story which can be self contained to the prover and the spec. That obviously would make the spec heavier which may not be a good thing.

Bottom line I am not sure what the right answer is, and I hope I have articulated my difficulty to come to term with the global error code stored on chain.

For the language we should all talk about it. I don't think global constant solve anything for the reason I mentioned. But if we think they are important for multiple reasons we can certainly talk about adding them. 
Also not sure who the language team is, and I'd like to think we are all part of it, so we need to put that into some kind of design conversation about what to do next. And global constants would require a binary format, VM changes so that specifically is not just the language.

As far as constants via functions go, I am with you. I hear you and understand. It's not a drama. Once again I'd love for us to be very minimalist on what we put on chain. And a module with a bunch of functions that return constants feels really bad to me. But again, I should really not let my taste control design and implementation decisions ;)

**Comment by dariorussi** [comment]
> I'm not super against the functions since if we allowed `public` constants, they'd have to probably compile to functions anyway. Otherwise they wouldn't make sense.
> 
> We could also allow constants to be defined out in the ether (outside of Modules or Script contexts). Which would basically make them a real fancy macro.

@tnowacki  I don't understand why public constants would have to be functions. And though this is not the right place to talk about design for constants :) I believe they can be defined as public constants (which is a constant with a public name). 
So I do not follow the "function" or "macro" story up there....

**Comment by dariorussi** [comment]
> Based on a discussion in a prover meeting today including Sam, I added a section about custom error codes to [Errors](https://github.com/wrwg/libra/blob/errors_module/language/stdlib/modules/Errors.move). It's at the end of the linked module. Please check out if you are wondering how to deal with those.
> 
> This also addresses this comment of @dariorussi:
> 
> > I am also very worried of us adding a module that contains a set of functions for constants for the current code base, which seems to imply that even the deployment of a new single and self containing module may require a WriteSet transaction to add an error.
> 
> The answer to this is that new error codes would be very rarely introduced in the `Errors` module, but as custom codes in the according module. A better name for `Errors` might be `CommonErrors`.

I suppose having a common errors could be an option and what you are saying is that those are truly heavily shared error codes that multiple modules may need. I guess if there are good reasons for that is hard to argue against!

**Comment by tnowacki** [comment]
> I believe they can be defined as public constants (which is a constant with a public name).

To be clear, I mean "they would behave in a manner similar to functions" in that they would have to be loaded dynamically. So if the feature was missing in the VM, we could compile them to functions and have essentially the same behavior

**Comment by tnowacki** [comment]
> I think the point that I am trying to make is that single repository of things in our system does not work too well. Whereas an approach where each module has a chance to be self contained has much less friction.

I think this is a good point. 
Consider if we have to writeset update one module. Is it worse with this single module that you now have to writeset two modules? It feels a bit more brittle, but I'm not sure if its more/less dangerous than a single-module writeset

**Comment by tzakian** [comment]
Still digesting the comments, but here are my initial thoughts on things: 

I'm overall a big fan of a centralized place for "core errors." I also think this has a benefit over a purely module-defined public constant solution (e.g., `LibraTimestamp` defining the `NOT_GENESIS` code). The central reason for this preference is due to the fact that a client can only tell the number and abort location, and not where the abort code was defined. In more detail: 
1. There is no way of tracking the abort code number back to the module that declared it; if you see a `MoveAbort{ code: 101, location:0x1::MyModule}` you don't know where you should start to look for the `101` error code. If we have a pre-specified range for "core errors" this means that there is only two modules that need to be looked at when resolving an abort code as a programmer; `MyModule` and `Errors` if the abort code is in that range. Additionally if we want to provide client documentation this means that every abort code needs to be unique in the system, which then means we should follow a similar approach to #4631..I would argue that this is in the long-term not sustainable.
2. The more centralized the error codes, the easier it is to make sure that codes are unique in their error range for these i.e., you don't need to worry about `MyModule`'s error codes stepping on top of `OtherModule`'s error codes.

In terms of what a centralized module for core error codes means for updating modules, and the additional friction this would cause; I agree that it is additional friction, but I disagree that this additional friction is onerous. Particularly I believe that it's not an undue burden because:
1. Updates to the core errors module can be performed asynchronously as long as the error is added before the module that uses it is added; and
2. These represent core/central errors in the system. I would expect the addition of such errors to the system to be very infrequent, and most likely to be batched.

**Comment by dariorussi** [comment]
> > I think the point that I am trying to make is that single repository of things in our system does not work too well. Whereas an approach where each module has a chance to be self contained has much less friction.
> 
> I think this is a good point.
> Consider if we have to writeset update one module. Is it worse with this single module that you now have to writeset two modules? It feels a bit more brittle, but I'm not sure if its more/less dangerous than a single-module writeset

yes you are right, though I was more thinking about a new module that would require no writeset normally, but with a central error code story would need a writeset

**Comment by wrwg** [comment]
> Some proposed rules of thumb for (1):
> A. Error codes are for consumption by clients, not for Libra framework developers or module authors.
> 
> B. Every error code should be considered part of the client-facing public API. If the error code doesn't make sense for someone who will read docs, but never look at framework code, it's too low-level. Similarly, every defined error code should be reachable from a current or planned future tx script (or else we should get rid of the error code).
> 
> C. Every error code should have an intended resolution. That is: the docs should say (or be able to say) "if you the client receive this error code, here is what you should do in your next transaction to avoid seeing this error code". If the error code doesn't have an intended resolution, it is too low-level and we should report the problem at a different level of abstraction. For example, instead of reporting a generic RESOURCE_NOT_INITIALIZED error code, we should emit an error code specific to the module/resource that tells the user which function/script/args they should use to initialize it.
> 
> D. An error code should be defined inside the module that understands how to resolve it. This is typically the same as the module that raises the error code for "leaf" modules (e.g., `LibraAccount::INSUFFICIENT_BALANCE`). But there are also "utility" modules like `Roles` that define many error codes + their resolutions, but do not raise the error codes.
> 
> E. A module `A` that `use`s another module `B` should reuse `B`'s error codes for problems that `B` knows how to resolve. E.g., `LibraAccount` should not define its own error code for `NOT_GENESIS`; it should reuse the `LibraTimestamp::NOT_GENESIS` error code.
> 
> Curious to hear which rules of thumb folks like/don't like and what other ones they have in mind.

I'm fully on the same page regards A and B but have trouble with D and E.

First, module abstractions are an implementation detail. Just recall that some of the DualAttestation logic was yesterday in LibraAccount and may be tomorrow somewhere else. The way how we structure modules inside of the framework is *not* part of the public API. This speaks in my mind against putting errors into modules *unless* they are rather specific to the functionality in this module (even then I would prefer to have them centralized because conceptually they are a big global enumeration and one needs to take care of keeping the codes disjoint, but I can compromise on this one).

Second, I think that an error would be "resolvable" locally to a module is an illusion in many cases. A good example for this would be `ALREADY_INITIALIZED`. This error happens if you call a function which publishes something during genesis twice. If the framework would tell a user `MINT_CAPABILITY_ALREADY_INITIALIZED` there would be little additional clue how to resolve this. If the user *is a developer* who wants to debug this, then what they would really need is a stack trace. It is 100x more valuable than the error code, how ever fine granular it is (it will never be granular enough compared to a stack trace).

The bottom line is if you think about error codes, you need to think about end users. For those, errors need to be translated into English and displayed as results of transactions. Many fine granular errors are not helpful to increase the users understanding. Something is "not found", "not allowed", "exceeds capacity", etc. is more than good. At the other end are users which develop on top of transaction scripts. For those a fine granular error regime sounds good, but in reality it is never good enough (see stack traces above). Therefore, I think it is not really worth the effort to come up with a detailed, fine granular classification of errors. Instead it increases the cognitive load on understanding and specifying the APIs.

File: language/stdlib/modules/LibraConfig.move [hunk 1]
```diff
@@ -92,9 +92,9 @@ module LibraConfig {
         config_account: &signer,
         payload: Config,
     ): ModifyConfigCapability<Config> {
-        assert(LibraTimestamp::is_genesis(), ENOT_GENESIS);
-        assert(Roles::has_libra_root_role(config_account), ENOT_LIBRA_ROOT);
-        assert(Signer::address_of(config_account) == CoreAddresses::LIBRA_ROOT_ADDRESS(), EINVALID_SINGLETON_ADDRESS);
+        assert(LibraTimestamp::is_genesis(), Errors::NOT_GENESIS());
```
[the remark this answers is not in the document]
Comment (wrwg) [reply language/stdlib/modules/LibraConfig.move:95]: Yes, that would need to be done, I thought in subsequent PR(s). It is ideally done together with a pass over the specs to clean them up and/or let them use other new features, like spec block level lets.

**Comment by wrwg** [comment]
Updates this PR by rebasing on the separately handled changes to invariants. Also renamed module `Errors` into `CoreErrors`.
